providers/freebsd/host_freebsd_cgo.go

```diff
@@ -0,0 +1,263 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//go:build freebsd && cgo
+
+package freebsd
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/elastic/go-sysinfo/internal/registry"
+	"github.com/elastic/go-sysinfo/providers/shared"
+	"github.com/elastic/go-sysinfo/types"
+)
+
+func init() {
+	registry.Register(newFreeBSDSystem())
+}
+
+type freebsdSystem struct{}
+
+func newFreeBSDSystem() freebsdSystem {
+	return freebsdSystem{}
+}
+
+func (s freebsdSystem) Host() (types.Host, error) {
+	return newHost()
+}
+
+type host struct {
+	procFS procFS
+	info   types.HostInfo
+}
+
+func (h *host) Info() types.HostInfo {
+	return h.info
+}
+
+func (h *host) CPUTime() (types.CPUTimes, error) {
+	cpu := types.CPUTimes{}
+	r := &reader{}
+	r.cpuTime(&cpu)
+	return cpu, r.Err()
+}
+
+func (h *host) Memory() (*types.HostMemoryInfo, error) {
+	m := &types.HostMemoryInfo{}
+	r := &reader{}
+	r.memInfo(m)
+	return m, r.Err()
+}
+
+func (h *host) FQDNWithContext(ctx context.Context) (string, error) {
+	return shared.FQDNWithContext(ctx)
+}
+
+func (h *host) FQDN() (string, error) {
+	return h.FQDNWithContext(context.Background())
+}
+
+func newHost() (*host, error) {
+	h := &host{}
+	r := &reader{}
+	r.architecture(h)
+	r.bootTime(h)
+	r.hostname(h)
+	r.network(h)
+	r.kernelVersion(h)
+	r.os(h)
+	r.time(h)
+	r.uniqueID(h)
+	return h, r.Err()
+}
+
+type reader struct {
+	errs []error
+}
+
+func (r *reader) addErr(err error) bool {
+	if err != nil {
+		if !errors.Is(err, types.ErrNotImplemented) {
+			r.errs = append(r.errs, err)
+		}
+		return true
+	}
+	return false
+}
+
+func (r *reader) Err() error {
+	if len(r.errs) > 0 {
+		return errors.Join(r.errs...)
+	}
+	return nil
```

Comment:
Apologies, but I don't understand the reasoning behind this change. The other providers are using [identical](https://github.com/elastic/go-sysinfo/blob/v1.14.0/providers/linux/host_linux.go#L185) methods.

Reply:
The reason is that it reduces the line count. `errors.Join` [returns nil if there is no non-nil error in the parameter list](https://pkg.go.dev/errors#Join). I suspect that the other cases look the way they do because the were written before `errors.Join` existed.